tests/test_ods.py

```diff
@@ -0,0 +1,214 @@
+import mock
+import functools
+
+import dmutils.ods as ods
+
+from hypothesis import strategies as st
+from hypothesis import given, example
+
+po = functools.partial(mock.patch.object, autospec=True)
+
+
+class TestRow(object):
+
+    @given(st.dictionaries(st.text(), st.text()))
+    def test___init__(self, kwargs):
+        with po(ods, 'TableRow') as TableRow:
+            instance = ods.Row(**kwargs)
+
+        TableRow.assert_called_once_with(**kwargs)
+
+        assert instance._row == TableRow.return_value
+
+    @given(st.text(), st.dictionaries(st.text(), st.text()))
+    @example("", {"numberrowsspanned": '1'})
+    @example("", {"numbercolumnsspanned": '1'})
+    def test_write_cell(self, value, kwargs):
+        instance = ods.Row()
+
+        instance._row = mock.MagicMock(spec_set=instance._row)
+
+        expected = dict(**kwargs)
+
+        if "numbercolumnsspanned" in kwargs:
+            expected.setdefault("numberrowsspanned", "1")
+
+        if "numberrowsspanned" in kwargs:
+            expected.setdefault("numbercolumnsspanned", "1")
+
+        ps = [mock.Mock() for v in value.split("\n")]
+
+        with po(ods, 'TableCell') as TableCell:
+            with po(ods, 'P') as P:
+                P.side_effect = iter(ps)
+                instance.write_cell(value, **kwargs)
+
+        TableCell.assert_called_once_with(**expected)
+
+        cell = TableCell.return_value
+        cell.setAttrNS.assert_called_once_with(ods.OFFICENS, 'value-type',
+                                               'string')
+
+        P.assert_has_calls([mock.call(text=line) for line in value.split("\n")])
+
+        cell.addElement.assert_has_calls([mock.call(e) for e in ps], True)
+
+        instance._row.addElement.assert_called_once_with(cell)
+
+    def test_write_covered_cell(self):
+        instance = ods.Row()
+
+        instance._row = mock.MagicMock(spec_set=instance._row)
+
+        with po(ods, 'CoveredTableCell') as CoveredTableCell:
+            instance.write_covered_cell()
+
+        cell = CoveredTableCell.return_value
+
+        instance._row.addElement.assert_called_once_with(cell)
+
+
+class TestSheet(object):
+    @given(st.text())
+    def test___init__(self, name):
+        with po(ods, 'Table') as Table:
```

Comment:
I'm a bit uncertain about unit testing that mocks implementation details. Sometimes you have to (e.g. API calls), but otherwise you're basically testing the implementation of the function rather than the end result, which makes for very fragile tests, that have to be rewritten if the implementation changes.

Not claiming I know how to do this better, but it's surely ok for a unit test to end up running code that's effectively at a lower layer than the code you're testing.

Reply:
test_ods was copied over from briefs, so I'm not really going to touch the tests here.

Reply:
ah fair enough then